pkg/snet/directtp/client.go

```diff
@@ -0,0 +1,468 @@
+package directtp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/AudriusButkevicius/pfilter"
+	"github.com/SkycoinProject/dmsg"
+	"github.com/SkycoinProject/dmsg/cipher"
+	"github.com/SkycoinProject/skycoin/src/util/logging"
+	"github.com/xtaci/kcp-go"
+
+	"github.com/SkycoinProject/skywire-mainnet/internal/packetfilter"
+	"github.com/SkycoinProject/skywire-mainnet/pkg/snet/arclient"
+	"github.com/SkycoinProject/skywire-mainnet/pkg/snet/directtp/pktable"
+	"github.com/SkycoinProject/skywire-mainnet/pkg/snet/directtp/porter"
+	"github.com/SkycoinProject/skywire-mainnet/pkg/snet/directtp/tpconn"
+	"github.com/SkycoinProject/skywire-mainnet/pkg/snet/directtp/tphandshake"
+	"github.com/SkycoinProject/skywire-mainnet/pkg/snet/directtp/tplistener"
+	"github.com/SkycoinProject/skywire-mainnet/pkg/snet/directtp/tptypes"
+)
+
+const (
+	// holePunchMessage is sent in a dummy UDP packet that is sent by both parties to establish UDP hole punching.
+	holePunchMessage   = "holepunch"
+	dialTimeout        = 30 * time.Second
+	dialConnPriority   = 20
+	visorsConnPriority = 100
+)
+
+var (
+	// ErrUnknownTransportType is returned when transport type is unknown.
+	ErrUnknownTransportType = errors.New("unknown transport type")
+
+	// ErrTimeout indicates a timeout.
+	ErrTimeout = errors.New("timeout")
+
+	// ErrAlreadyListening is returned when transport is already listening.
+	ErrAlreadyListening = errors.New("already listening")
+
+	// ErrPortOccupied is returned when port is occupied.
+	ErrPortOccupied = errors.New("port is already occupied")
+)
+
+// Client is the central control for incoming and outgoing 'Conn's.
+type Client interface {
+	Dial(ctx context.Context, rPK cipher.PubKey, rPort uint16) (*tpconn.Conn, error)
+	Listen(lPort uint16) (*tplistener.Listener, error)
+	Serve() error
+	Close() error
+	Type() string
+}
+
+// Config configures Client.
+type Config struct {
+	Type            string
+	PK              cipher.PubKey
+	SK              cipher.SecKey
+	LocalAddr       string
+	Table           pktable.PKTable
+	AddressResolver arclient.APIClient
+}
+
+type client struct {
+	conf           Config
+	mu             sync.Mutex
+	done           chan struct{}
+	once           sync.Once
+	log            *logging.Logger
+	porter         *porter.Porter
+	packetFilter   *pfilter.PacketFilter
+	packetListener net.PacketConn
```

Comment:
we have `packetListener` and `listener`. if I understand correctly, first one is for SUDPH, second one is for STCP/R tps. what do you think about reflecting this in the names of these fields? just asking, probably it'll make this clearer

Reply:
Good idea